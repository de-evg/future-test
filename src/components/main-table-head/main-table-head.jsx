import React, { useCallback } from "react";
import PropTypes from "prop-types";
import { NameSpace } from "../../store/reducers/root";
import { connect } from "react-redux";
import { ActionCreator } from "../../store/action";

const tableHeads = [`ID`, `First Name`, `Last Name`, `Email`, `Phone`];
const HeadValues = {
  [`ID`]: `ID`,
  [`First Name`]: `FIRSTNAME`,
  [`Last Name`]: `LASTNAME`,
  [`Email`]: `EMAIL`,
  [`Phone`]: `PHONE`
};

const SortName = {
  [`ID`]: (activeSort) => activeSort === `ID_UP` ? `ID_DOWN` : `ID_UP`,
  [`First Name`]: (activeSort) =>  activeSort === `FIRSTNAME_UP` ? `FIRSTNAME_DOWN` : `FIRSTNAME_UP`,
  [`Last Name`]: (activeSort) =>  activeSort === `LASTNAME_UP` ? `LASTNAME_DOWN` : `LASTNAME_UP`, 
  [`Email`]: (activeSort) => activeSort === `EMAIL_UP` ? `EMAIL_DOWN` : `EMAIL_UP`,
  [`Phone`]: (activeSort) => activeSort === `PHONE_UP` ? `PHONE_DOWN` : `PHONE_UP`
};

const MainTableHead = ({ activeSort, updateSort }) => {
  const handleClick = useCallback(
    (evt) => {
      updateSort(SortName[evt.target.dataset.sort](activeSort));
    },
    [updateSort, activeSort]
  );
    const [sortType, sortDirection] = activeSort.split("_");
  return (
    <thead>
      <tr className="table__header table__row">
        {tableHeads.map((head, i) => (
          <th
            key={`head-${i}`}            
            data-sort={head}
            className="table__column-title table__data"
            onClick={handleClick}
          >
            {head}
            {HeadValues[head] === sortType && sortDirection === `UP` && (
              <span data-sort={head} className="sort__arrow-up"></span>
            )}
            {HeadValues[head] === sortType && sortDirection === `DOWN` && (
              <span data-sort={head} className="sort__arrow-down"></span>
            )}
          </th>
        ))}
      </tr>
    </thead>
  );
};

MainTableHead.propTypes = {
  activeSort: PropTypes.string.isRequired,
  updateSort: PropTypes.func.isRequired
};

const mapStateToProps = (state) => ({
  activeSort: state[NameSpace.USERS_DATA].activeSort,
});

const mapDispatchToProps = (dispatch) => ({
  updateSort(sort) {
    dispatch(ActionCreator.updateSort(sort));
  },
});
export {MainTableHead};
export default connect(mapStateToProps, mapDispatchToProps)(MainTableHead);
