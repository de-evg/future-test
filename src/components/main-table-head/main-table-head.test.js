import React from 'react';
import Enzyme, {shallow} from 'enzyme';
import {shallowToJson} from 'enzyme-to-json';
import {MainTableHead} from './main-table-head';
import Adapter from "enzyme-adapter-react-16";

Enzyme.configure({adapter: new Adapter()});

it(`Render MainTableHead correctly`, () => {
  const wrapper = shallow(<MainTableHead activeSort={`head_UP`} updateSort={() => {}} />)
  const tree = shallowToJson(wrapper);
  expect(tree).toMatchSnapshot();
});