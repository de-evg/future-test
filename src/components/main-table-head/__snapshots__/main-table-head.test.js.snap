// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render MainTableHead correctly 1`] = `
<thead>
  <tr
    className="table__header table__row"
  >
    <th
      className="table__column-title table__data"
      data-sort="ID"
      key="head-0"
      onClick={[Function]}
    >
      ID
    </th>
    <th
      className="table__column-title table__data"
      data-sort="First Name"
      key="head-1"
      onClick={[Function]}
    >
      First Name
    </th>
    <th
      className="table__column-title table__data"
      data-sort="Last Name"
      key="head-2"
      onClick={[Function]}
    >
      Last Name
    </th>
    <th
      className="table__column-title table__data"
      data-sort="Email"
      key="head-3"
      onClick={[Function]}
    >
      Email
    </th>
    <th
      className="table__column-title table__data"
      data-sort="Phone"
      key="head-4"
      onClick={[Function]}
    >
      Phone
    </th>
  </tr>
</thead>
`;
